### PR TITLE
GEODE-2109 : Calling submit on ExecutionService can cause exceptions to be lost

### PR DESCRIPTION
**Pull request by deepakddixit**
Replaced submit call with execute.
Refactored GIITask calls from DiskStoreImpl. Async task like Creating KRF's through disk store are replaced with execute. The other call is kept with submit as we later check for whether delayed write is completed or not.
Added RejectedExecutionException to IgnoredException list as from these tests it may get occasionally thrown and as we moved to execute method of ExecutorService they will be logged.

All precheckin tests are passing except one org.apache.geode.management.internal.cli.commands.IndexCommandsDUnitTest.testCreateDestroyUpdatesSharedConfig which is marked as Flaky.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java [hunk 1]
```diff
@@ -386,7 +398,7 @@ public void startManagingActivity() throws Exception {
    * 
    */
 
-  private class GIITask implements Callable<DistributedMember> {
+  private class GIITask implements Callable {
```
Comment (upthewaterspout) [inline geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java]: Why did you remove <DistributedMember> from this class declaration?

**Comment by deepakddixit** [comment]
@upthewaterspout Thanks for the review. The class declaration is mistakenly changed. I have corrected it. All tests are passing from precheckin after this.

**Comment by kirklund** [comment]
We've been requiring unit tests or integration tests for all bug fixes. 

At a minimum, I would recommend adding a new test that fails due to at least of the conditions reported by GEODE-2109 and then passes with your changes. Example:
`
import static org.assertj.core.api.Assertions.assertThat; 

import org.apache.geode.test.junit.categories.UnitTest;
import org.junit.Rule;
import org.junit.Test;
import org.junit.contrib.java.lang.system.SystemErrRule;
import org.junit.experimental.categories.Category;

@Category(UnitTest.class)
public class FederatingManagerSubmitTaskWithFailureTest {

  private FederatingManager manager;

  @Rule
  public SystemErrRule systemErrRule = new SystemErrRule().enableLog();

  @Test
  public void submittedTaskShouldLogFailure() throws Exception {
    manager.submitTask(() -> {throw new Exception("error message");});
    assertThat(systemErrRule.getLog()).contains(Exception.class.getName()).contains("error message");
  }
}
`
The above test would require a couple things:
1) some sort of @Before method which instantiates manager
2) change FederatingManager.submitTask(...) to package scope so the unit test can call it (right now FederatingManager is not a test-friendly class)

**Comment by deepakddixit** [comment]
@kirklund I have added a test as suggested with SingleHopTestExecutor. Kindly review.

**Comment by kirklund** [comment]
I think this is ready to commit. I'll check with @upthewaterspout. Thanks!

**Comment by upthewaterspout** [comment]
+1 looks good to me.

**Comment by upthewaterspout** [comment]
I looking at pulling this, and I see what one of the unit tests failed for me. It's failing intermittently in the IDE:
    SingleHopClientExecutorSubmitTaskWithExceptionTest.submittedTaskShouldLogFailure

    java.lang.RuntimeException: I am expecting this to be logged
	at  org.apache.geode.cache.client.internal.SingleHopClientExecutorSubmitTaskWithExceptionTest$1.run(SingleHopClientExecutorSubmitTaskWithExceptionTest.java:52)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

**Comment by deepakddixit** [comment]
@upthewaterspout It looks like from IDE when we run all tests under UnitTest category above test fails. But when we run either :geode-core:test target or complete precheckin I never observed such failure. I have observed that in IDE, SystemErrRule.getLog is not populated from thrown exception.

**Comment by upthewaterspout** [comment]
@deepakddixit I was running ./gradlew test when it failed. What's weird to me is that we're not seeing the stack trace from the test for what actually failed the test, but I think there's definitely something doing wrong here. Maybe it is a bug in SystemErrRule.

**Comment by deepakddixit** [comment]
Precheckin and ./gradlew test passed on my machine.  @upthewaterspout Kindly have a look over updated PR.

**Comment by deepakddixit** [comment]
@upthewaterspout Do changes look good?

**Comment by upthewaterspout** [comment]
Hi @deepakddixit 

Sorry for the slow response. I'm a little concerned that with your changes to the test in ef4d38d, the test is no longer actually testing what happens if the runnable passed to the executor throws an exception.

I think those changes *do* show that the problem is not with SystemErrRule, since with your change the test passes consistently. What that means is that there might be a product issue.

I did a little bit of investigation, and I found that if I set an uncaught exception handler in the product, your original test passes consistently. I'm not quite sure why the default uncaught exception handler was not consistently working, but I think it's better to send the error to the log anyway.

I stuck my changes up as a pull request to your branch if want to look at them. Let me know what you think, if this looks good to you I can merge this stuff:

https://github.com/deepakddixit/incubator-geode/pull/1

**Comment by deepakddixit** [comment]
@upthewaterspout Thanks for the detailed review. I agree with you, SingleHopClientExecutor was not having uncaught exception handler which is causing exceptions not handled and ultimately not logged. Setting an uncaughtExceptionHandler would help in logging uncaught exceptions. 
Only change I would like to suggest is instead of adding uncaughtExceptionHandler can we set LoggingThreadGroup? The other pooled executors are using LoggingThreadGroup which handles uncauhtExceptions. I have changed code a bit and added LoggingThreadGroup, unit tests are passing. Let me know your view on using LoggingThreadGroup.

**Comment by upthewaterspout** [comment]
@deepakddixit - I think using LoggingThreadGroup is definitely a better fix. The changes look good to me now, I'll merge them. Thanks!

**Comment by deepakddixit** [comment]
Thanks @upthewaterspout for your assistance in resolving it.
